### DISCUSSION TITLE
Removing multiple experiments at once

### DISCUSSION THREAD

**Discussion by Did-Mas** [pymeasure/pymeasure#1189] [Ideas]
If I am not mistaken, currently there is no way to remove multiple experiments (but not all) from the browser at once. For example, if a user wants to remove all queued experiments, they have to right-click each one and select "Remove graph" or "Delete data file."

I think an option to clear experiments by status would be useful. This could be implemented by adding a "Clear by status" button to the ```BrowserWidget``` next to the "Clear all" button. When pressed, the button could open a simple dialog allowing the user to select which statuses should be removed and whether the corresponding files should be deleted.

Added button:
![image](https://github.com/user-attachments/assets/7010fec7-c486-416b-a3a5-8341f6dccf08)

Dialog:
![image](https://github.com/user-attachments/assets/60a8868a-211b-4b9e-952a-0dabd55a2beb)

Additionally, it could be useful to have the option to remove only selected files, such as those currently visible. However, I don’t have a clear idea of how to implement this without cluttering the GUI too much.

**Comment by CasperSchippers** [comment]
Hey, You are right, and there is no such option at the moment, but I like the idea. Not sure I'm like the dialog to be honest though, but not cluttering the GUI is indeed a good idea and quite a difficult challenge always.
I've been giving it some though and, by chance ran into an option that I quite like: apparently you can add a menu to a button:
```python
self.delete_menu = QtWidgets.QMenu(self)
self.delete_menu.addAction("all")
self.delete_menu.addAction("finished")
self.delete_menu.addAction("aborted")
self.delete_menu.addAction("failed")
self.delete_menu.addAction("selected")

self.delete_menu_button = QtWidgets.QPushButton("Clear ... ", self)
self.delete_menu_button.setMenu(self.delete_menu)
```
resulting in:
![image](https://github.com/user-attachments/assets/15e226af-a9a7-4402-bb75-f163140b9d5e)
![image](https://github.com/user-attachments/assets/4b236cff-0a0e-4d8e-92d4-177200998461)

This, of course, is a very simple example, and does not do anything, and are not enabled/disabled like the other buttons, but that should all be possible. And it also allows to put more stuf in there (with either a more general dropdown menu button thing, or adjacent menus for other groups) without cluttering the interface.

What do you think?

**Reply by Did-Mas** [reply]
Oh, I wasn’t aware you could add a menu to a button—that’s a great idea! I agree it’s cleaner and more visually appealing than a dialog, and it might even serve as a replacement for the "Clear All" button.

One aspect I’m uncertain about is how to differentiate between deleting a file entirely and simply removing it from the app. Perhaps two separate buttons could work, but then again the clutter.

I also noticed an inconsistency in how the functions are named. The "Clear All" button removes files from the app, but the equivalent functionality for a single file in the browser item context menu is labeled "Remove graph." This naming could be a bit counterintuitive.

**Reply by CasperSchippers** [reply]
Yeah, me neither, just came across it by chance when looking for something else entirely.

That is indeed a good question. One option I can think of is putting everything in a single menu, either in sections within that menu (you can add separators or named "sections" using disabled actions maybe) or in a nested menu of some form (so a delete button/menu, with either 2 sub menus (delete w/ or w/o file), or maybe the other way around (2 options for every of the selections).
Another option is via a confirmation dialog, to which you can add a checkbox for deleting the file. I'm not sure what my preference would be. Anyways, when also deleting the file, I tend to like an "are you sure" dialog, but for only clearing the measurement list, it's not that important.

Regarding the naming, consistency would be nice indeed. I think it's fine to rename them in favor of clarity; I don't think a cosmetic change is that much of a deal.

**Reply by Benbriel** [reply]
A checkbox can be added to the menu, allowing users to choose whether to include files when removing graphs. Sadly, clicking the checkbox closes the menu, so it needs to be reopened manually afterward. Here's an example:

```python
self.delete_includes_files = self.delete_menu.addAction("Include Files")
self.delete_includes_files.setCheckable(True)
self.delete_includes_files.triggered.connect(self.delete_menu.show)
self.delete_menu.addSeparator()
```
![image](https://github.com/user-attachments/assets/8fbcd8d2-3dfd-4fa6-a6b7-76a3a30d430f)

**Reply by CasperSchippers** [reply]
Hmm, that is maybe quite a nice solution, nice find! Too bad of the closing, but oh well. I´d suggest going this route!!
